processor/src/main/java/org/mapstruct/ap/internal/util/SpecificCompilerWorkarounds.java

```diff
@@ -115,4 +135,161 @@ public static TypeElement replaceTypeElementIfNecessary(Elements elementUtils, T
         }
         return element;
     }
+
+    public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                        Elements elementUtils) {
+        try {
+            // quick and dirty, until implementation below is more sound
+            return typeUtils.asMemberOf( containing, element );
+        }
+        catch ( IllegalArgumentException e ) {
+            TypeMirror eclipseWorkaround =
+                EclipseEnvironment.asMemberOf( containing, element, typeUtils, elementUtils );
+
+            if ( null != eclipseWorkaround ) {
+                return eclipseWorkaround;
+            }
+
+            throw e;
+        }
+    }
+
+    private static final class EclipseEnvironment {
```

Comment:
Are you sure this class *only* is loaded in the Eclipse case? All of bit being in one class makes me a bit nervous. I know it'll be separate class files, but still...